File: ass2/adt/disjoint_set.py
#!/usr/bin/env python3
# pyright: basic

from __future__ import annotations


class DisjointSet:
    def __init__(self):
        self.parent: dict[int, int] = {}
        self.rank: dict[int, int] = {}

    def make_set(self, x: int):
        if x not in self.parent:
            self.parent[x] = x
            self.rank[x] = 0

    def find_set(self, x: int) -> int:
        # Path compression
        if self.parent[x] != x:
            self.parent[x] = self.find_set(self.parent[x])
        return self.parent[x]

    def union(self, x: int, y: int):
        rx, ry = self.find_set(x), self.find_set(y)
        if rx == ry:
            return

        # Union by rank
        if self.rank[rx] < self.rank[ry]:
            self.parent[rx] = ry
        elif self.rank[rx] > self.rank[ry]:
            self.parent[ry] = rx
        else:
            self.parent[ry] = rx
            self.rank[rx] += 1

File: ass3/tasks/disjoint_set.go
package tasks

// Go port of ass2/adt/disjoint_set.py

type DisjointSet struct {
	parent map[string]string
	rank   map[string]int
}

func newDisjointSet() *DisjointSet {
	return &DisjointSet{make(map[string]string), make(map[string]int)}
}

func (ds *DisjointSet) makeSet(x string) {
	if _, ok := ds.parent[x]; !ok {
		ds.parent[x] = x
		ds.rank[x] = 0
	}
}

func (ds *DisjointSet) findSet(x string) string {
	if ds.parent[x] != x {
		ds.parent[x] = ds.findSet(ds.parent[x])
	}
	return ds.parent[x]
}

func (ds *DisjointSet) union(x, y string) {
	rx, ry := ds.findSet(x), ds.findSet(y)
	if rx == ry {
		return
	}
	if ds.rank[rx] < ds.rank[ry] {
		ds.parent[rx] = ry
	} else if ds.rank[rx] > ds.rank[ry] {
		ds.parent[ry] = rx
	} else {
		ds.parent[ry] = rx
		ds.rank[rx]++
	}
}
